### PR TITLE
Add build support for arm64 architecture

### PR DESCRIPTION
**Pull request by popegeo**
Check the main Makefile to suport the ARM64 architecture, considering the
default build architecture x86_64.
Building on arm64 on a different architecture requires a toolchain
to be installed by the user. The default toolchain chosen is
the one offered by Linaro. Also, a specific toolchain can be set by
defining the CROSS_COMPILE environment varriable.

An example of building on arm64 architecture is:

make ARCH=arm64

Future commits will move the build inside a Docker container where the
default toolchain will be automatically installed.

Signed-off-by: George-Aurelian Popescu <popegeo@amazon.com>

*Observation*
I don't know why git displays the Makefile formatting like that, but if you open the Makefile from the repository you'll see that
are the variables are aligned. 
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by petreeftime** [comment]
> Observation
I don't know why git displays the Makefile formatting like that, but if you open the Makefile from the repository you'll see that
are the variables are aligned.

It's likely because tabs and spaces are used on the same line. I would suggest using just spaces for in-line aligning.

File: Makefile [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
+ARCH_x86_64 	       = x86_64
+ARCH_aarch64	       = arm64
+KERN_IMAGE_x86_64      = bzImage
+KERN_IMAGE_aarch64     = Image
+CONFIG_x86_64	       = microvm-kernel-config-x86_64
+CONFIG_aarch64	       = microvm-kernel-config-aarch64
+DEFAULT_CROSS_COMPILER = /usr/bin/aarch64-linux-gnu-
+
+ifndef ARCH
+	ARCH=$(ARCH_x86_64)
```
Comment (petreeftime) [inline Makefile:15]: I think using the local arch would be recommended instead of using x86_64 by default.

File: Makefile [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+endif
+
+ifeq ($(ARCH),$(ARCH_x86_64))
+	KERN_IMAGE=$(KERN_IMAGE_x86_64)
+	CONFIG=configs/$(CONFIG_x86_64)
+else ifeq ($(ARCH),$(ARCH_aarch64))
+	KERN_IMAGE=$(KERN_IMAGE_aarch64)
+	CONFIG=configs/$(CONFIG_aarch64)
+	ifndef CROSS_COMPILE
+		CROSS_COMPILE=$(DEFAULT_CROSS_COMPILER)
```
Comment (petreeftime) [inline Makefile:25]: Use the CROSS_COMPILE toolchain only if cross-compiling.

File: Makefile [hunk 3]
```diff
@@ -1,12 +1,34 @@
...
-OUTDIR   ?= $(abspath blobs)
+BUILDDIR 	      ?= $(abspath build)
+OUTDIR		      ?= $(abspath blobs)
+ARCH_x86_64 	       = x86_64
+ARCH_aarch64	       = arm64
+KERN_IMAGE_x86_64      = bzImage
+KERN_IMAGE_aarch64     = Image
+CONFIG_x86_64	       = microvm-kernel-config-x86_64
+CONFIG_aarch64	       = microvm-kernel-config-aarch64
+DEFAULT_CROSS_COMPILER = /usr/bin/aarch64-linux-gnu-
```
Comment (petreeftime) [inline Makefile:12]: DEFAULT_AARCH64_CROSS_COMPILER / DEFAULT_X86_64_CROSS_COMPILER.

File: Makefile [hunk 4]
```diff
@@ -1,12 +1,36 @@
...
+
+CONFIG=$(CONFIG_TEMPLATE)$(BUILD_ARCH)
+
+ifeq ($(BUILD_ARCH),$(ARCH_x86_64))
+	KERN_IMAGE=$(KERN_IMAGE_x86_64)
+	ARCH=$(ARCH_x86_64)
+else ifeq ($(BUILD_ARCH),$(ARCH_aarch64))
+	KERN_IMAGE=$(KERN_IMAGE_aarch64)
+	ifndef CROSS_COMPILE
+		ifneq ($(ARCH),$(HOST_MACHINE))
```
Comment (andraprs) [inline Makefile:24]: Shouldn't here be BUILD_ARCH instead of ARCH?

Comment (popegeo) [reply Makefile:24]: yes it should, thanks for the notice

File: init/Makefile [hunk 5]
```diff
@@ -1,11 +1,20 @@
-CC       = gcc
-C_FLAGS := -Wall -Wextra -Werror -O2
+
+ARCH_aarch64	       = arm64
+C_FLAGS 	      := -Wall -Wextra -Werror -O2
+CC 		      := gcc
+DEFAULT_CROSS_COMPILER = /usr/bin/aarch64-linux-gnu-
+
+ifeq ($(ARCH),$(ARCH_aarch64))
```
Comment (andraprs) [inline init/Makefile:7]: In the main Makefile, the one from above, ARCH should be passed in "make -C $(PWD)/init" via ARCH=$(ARCH).

And should update here as well, to check for the HOST_MACHINE arch and set cross compile if it is not Arm64.

Comment (popegeo) [reply init/Makefile:7]: yes, I am working on a fixup there too :D

**Review by andraprs — DISMISSED** [review]

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+RUN mkdir /build
+
+ADD ./ /build/
+
+WORKDIR /build
+
+RUN gpg2 --locate-keys torvalds@kernel.org gregkh@kernel.org -vv
+RUN make BUILD_ARCH=${BUILD_ARCH}
+
```
Comment (andraprs) [inline Dockerfile:19]: There are a couple blank lines below.

File: init/Makefile [hunk 7]
```diff
@@ -1,11 +1,25 @@
-CC       = gcc
-C_FLAGS := -Wall -Wextra -Werror -O2
+
+ARCH_aarch64	               = aarch64
+C_FLAGS 	              := -Wall -Wextra -Werror -O2
```
Comment (andraprs) [inline init/Makefile:3]: These 2 (C_FLAGS and CC) should be aligned with the other ones; if not already, and it's only viewing like this on GitHub.

**Review by andraprs — APPROVED** [review]

File: init/Makefile (hunk 7 above)
Comment (andraprs) [reply init/Makefile]: Looks ok now.

**Review by laurstefan — APPROVED** [review]

**Pull request merged**